### DISCUSSION TITLE
Python 3 on Big Sur

### DISCUSSION THREAD

**Discussion by johnalanwoods** [Homebrew/discussions#320] [Getting started] [answered]
Hi, 

I've installed Python3.9 using brew on Big Sur.
It notes after install:

```
Unversioned symlinks `python`, `python-config`, `pip` etc. pointing to
`python3`, `python3-config`, `pip3` etc., respectively, have been installed into
  /opt/homebrew/opt/python@3.9/libexec/bin
```

I used to just do `export PATH="/opt/homebrew/opt/python@3.9/libexec/bin:$PATH"`

but I noticed today that as part of Xcode, Python3 is already installed at `/usr/bin/python3`.

Even though I have `/opt/homebrew/bin` in my Path, Python3 is being picked up at `/usr/bin/` first.

```
➜  ~ echo $PATH
/opt/homebrew/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin
```

```
➜  ~ which python3
/usr/bin/python3
```

```
➜  ~ type -a python3
python3 is /opt/homebrew/bin/python3
python3 is /usr/bin/python3
```

What's the right way to have it so calling "python" at the command line uses brew, but I don't screw up the system python?

**Comment by gromgit** [comment, marked as answer]
The system python3 reference may have been cached in your current shell session. Assuming you're using bash, try `hash -r`, then `which python3` again.

**Reply by johnalanwoods** [reply]
@gromgit - OK that was really stupid, but in my defence, I had opened a new tab/shell a number of times, in the end I had to restart iTerm2 for it to work.

Always thought a new tab/shell was enough!

**Reply by johnalanwoods** [reply]
@gromgit, I guess since python is installed to `/opt/homebrew/bin/python3`, I don't need to `export PATH="/opt/homebrew/opt/python@3.9/libexec/bin:$PATH"` - right?

**Reply by carlocab** [reply]
You don't need to. Doing that allows you to call Homebrew `python3` as `python` (and `pip3` as `pip`, etc) in your shell, however.

**Reply by gromgit** [reply]
> Always thought a new tab/shell was enough!

It should be, actually, since bash caches aren't shared across instances. Sounds like your shell startup files somehow invoke `python3` before `/opt/homerew/bin` is added to `PATH`, hence `bash` adds the system Python to its cache.

**Reply by infa-jowoods** [reply]
Thanks guys!

**Comment by psirenko** [comment]
Hello! 
I use macOS 11.1 + Homebrew 2.6.2. I've installed python@3.9 and received the note:

```
Python has been installed as
  /usr/local/bin/python3

Unversioned symlinks `python`, `python-config`, `pip` etc. pointing to
`python3`, `python3-config`, `pip3` etc., respectively, have been installed into
  /usr/local/opt/python@3.9/libexec/bin

You can install Python packages with
  pip3 install <package>
They will install into the site-package directory
  /usr/local/lib/python3.9/site-packages

See: https://docs.brew.sh/Homebrew-and-Python
==> Summary
🍺  /usr/local/Cellar/python@3.9/3.9.1_1: 3,895 files, 63.9MB
```

I don't see the IDLE in my application catalog. What should I do to fix this?

**Reply by gromgit** [reply]
You should have an `idle3` symlink in `/usr/local/bin`.

Tip: To see what `<formula>` contains, run `brew ls <formula>`

**Reply by psirenko** [reply]
@gromgit Hello! Could you please help me to create this symlink?
I used `brew link --overwrite python`

```
petr@air-anastasia ~ % brew link --overwrite python
Warning: Already linked: /usr/local/Cellar/python@3.9/3.9.1_1
To relink:
  brew unlink python@3.9 && brew link python@3.9
petr@air-anastasia ~ % which python
/usr/bin/python
petr@air-anastasia ~ % which idle3
/usr/local/bin/idle3
```

**Reply by carlocab** [reply]
~You don't need to create it. It should've been created for you upon installing python.~

Right, you edited your post after I wrote the above...

**Reply by psirenko** [reply]
> You don't need to create it. It should've been created for you upon installing python.
Thank you! I agree with you, but something went wrong and I can't use idle now :(

I have also created an issue in [Python Bug tracker](https://bugs.python.org/issue42691). Maybe somebody will help me to understand what is wrong and what to do

**Reply by gromgit** [reply]
I'm only a Python dabbler, but I get the impression that general "best practice" is to assume that `python` gets you Python *2*, not 3, and that `python3` is the "blessed" way to get the latter. That's likely why:
```
Unversioned symlinks `python`, `python-config`, `pip` etc. pointing to
`python3`, `python3-config`, `pip3` etc., respectively, have been installed into
  /usr/local/opt/python@3.9/libexec/bin
```
to avoid unwanted confusion.

**Reply by alebcay** [reply]
[PEP 394](https://www.python.org/dev/peps/pep-0394/) may be helpful in understanding the rationale behind the naming convention (both historically and currently).

**Comment by ShenYj** [comment]
```zsh
╰─ brew config                                                                                                   ─╯
HOMEBREW_VERSION: 2.7.4
ORIGIN: https://github.com/Homebrew/brew
HEAD: 663b972f449b679512a2c0e3303569a2e9bfba7c
Last commit: 31 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core.git
Core tap HEAD: 67738bee94d8433a96091dc5848aea3b218f3766
Core tap last commit: 70 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: :0
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: dodeca-core 64-bit kabylake
Clang: 12.0 build 1200
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.1-x86_64
CLT: 12.3.0.0.1.1607026830
Xcode: 12.3
```

```zsh
╰─ which python3                                                                                                 ─╯
/usr/local/bin/python3
```

get the same problem with:
```zsh
==> python@3.9
Python has been installed as
  /usr/local/bin/python3

Unversioned symlinks `python`, `python-config`, `pip` etc. pointing to
`python3`, `python3-config`, `pip3` etc., respectively, have been installed into
  /usr/local/opt/python@3.9/libexec/bin

You can install Python packages with
  pip3 install <package>
They will install into the site-package directory
  /usr/local/lib/python3.9/site-packages

See: https://docs.brew.sh/Homebrew-and-Python
```

**Reply by carlocab** [reply]
What's the problem here?

```
╰─ which python3                                                                                                 ─╯
/usr/local/bin/python3
```
This says that calling `python3` in your shell gives you Homebrew `python3`.

**Comment by roncromberge** [comment]
Stupid question,

Why can't I delete the /usr/bin/python3 (3.8...) version? (with sudo of course )

I don't like dangling things in my system :-)

**Reply by SMillerDev** [reply]
Because that's part of your OS and SIP won't allow it.
